File: backend/app/controllers/member_controller.js
import { DataService } from '../services/data_service.js';

export class MemberController {
  constructor(churnService) {
    this.churnService = churnService;
  }

  getMembers = (req, res) => {
    try {
      const { members, schema } = DataService.loadDataset();
      const { search, riskLevel, planType, page = '1', limit = '20', sortBy = 'churnProbability', sortOrder = 'desc' } = req.query;

      let result = members.map(m => {
        const { probability, riskLevel: risk, prediction } = this.churnService.predictMember(m);
        return {
          ...m,
          churnProbability: probability,
          riskLevel: risk,
          prediction,
        };
      });

      // Filter search
      if (search) {
        const q = String(search).toLowerCase();
        result = result.filter(m => 
          String(m[schema.idColumn]).toLowerCase().includes(q) ||
          String(m.Plan_Type || '').toLowerCase().includes(q) ||
          String(m.Plan_Tier || '').toLowerCase().includes(q)
        );
      }

      // Filter risk
      if (riskLevel && riskLevel !== 'ALL') {
        result = result.filter(m => m.riskLevel === String(riskLevel).toUpperCase());
      }

      // Filter plan
      if (planType && planType !== 'ALL') {
        result = result.filter(m => String(m.Plan_Type).toLowerCase() === String(planType).toLowerCase());
      }

      // Sort
      result.sort((a, b) => {
        let valA = a[String(sortBy)] ?? 0;
        let valB = b[String(sortBy)] ?? 0;
        if (typeof valA === 'string') valA = valA.toLowerCase();
        if (typeof valB === 'string') valB = valB.toLowerCase();

        if (valA < valB) return sortOrder === 'asc' ? -1 : 1;
        if (valA > valB) return sortOrder === 'asc' ? 1 : -1;
        return 0;
      });

      // Pagination
      const pageNum = parseInt(String(page), 10) || 1;
      const limitNum = parseInt(String(limit), 10) || 20;
      const totalCount = result.length;
      const startIndex = (pageNum - 1) * limitNum;
      const paginatedMembers = result.slice(startIndex, startIndex + limitNum);

      res.json({
        totalCount,
        page: pageNum,
        limit: limitNum,
        totalPages: Math.ceil(totalCount / limitNum),
        members: paginatedMembers,
      });
    } catch (err) {
      res.status(500).json({ error: err.message || 'Error fetching members' });
    }
  };

  getMemberById = async (req, res) => {
    try {
      const memberId = req.params.member_id;
      const member = DataService.getMemberById(memberId);
      if (!member) {
        return res.status(404).json({ error: `Member ${memberId} not found` });
      }

      const { members, schema } = DataService.loadDataset();
      const processed = this.churnService.getProcessedDataset();
      
      const { probability, riskLevel, prediction } = this.churnService.predictMember(member);
      
      const { ExplainabilityService } = await import('../services/explainability_service.js');
      const { RetentionService } = await import('../services/retention_service.js');
      const { GeminiService } = await import('../services/gemini_service.js');

      const topDrivers = processed 
        ? ExplainabilityService.getMemberDrivers(member, schema, processed, this.churnService, 5)
        : [];

      const recommendations = RetentionService.generateRecommendations(member, riskLevel, probability, topDrivers);

      const aiExplanation = await GeminiService.generateMemberExplanation(
        String(member[schema.idColumn]),
        String(member.Plan_Type || 'Plan'),
        probability,
        riskLevel,
        topDrivers,
        recommendations
      );

      res.json({
        member,
        churnProbability: probability,
        riskLevel,
        prediction,
        topDrivers,
        recommendations,
        aiExplanation,
      });
    } catch (err) {
      res.status(500).json({ error: err.message || 'Error fetching member details' });
    }
  };
}

File: backend/app/controllers/member_controller.py
import math
from typing import Dict, Any, Optional
from fastapi import HTTPException
from app.services.data_service import DataService
from app.services.explainability_service import ExplainabilityService
from app.services.retention_service import RetentionService
from app.services.gemini_service import GeminiService

class MemberController:
    def __init__(self, churn_service):
        self.churn_service = churn_service

    def get_members(
        self,
        search: Optional[str] = None,
        riskLevel: Optional[str] = None,
        planType: Optional[str] = None,
        page: int = 1,
        limit: int = 20,
        sortBy: str = "churnProbability",
        sortOrder: str = "desc"
    ) -> Dict[str, Any]:
        members, schema = DataService.load_dataset()
        id_col = schema["idColumn"]

        result = []
        for i, m in enumerate(members):
            m_id = str(m.get(id_col) or m.get("Member_ID", f"MMB-{10001 + i}"))
            pred_res = self.churn_service.get_member_prediction_fast(m_id, m)
            item = dict(m)
            item["churnProbability"] = pred_res["probability"]
            item["riskLevel"] = pred_res["riskLevel"]
            item["prediction"] = pred_res["prediction"]
            result.append(item)

        # Filter search
        if search:
            q = str(search).lower()
            filtered = []
            for m in result:
                id_val = str(m.get(id_col, "")).lower()
                plan_t = str(m.get("Plan_Type", "")).lower()
                plan_tier = str(m.get("Plan_Tier", "")).lower()
                if q in id_val or q in plan_t or q in plan_tier:
                    filtered.append(m)
            result = filtered

        # Filter risk
        if riskLevel and riskLevel.upper() != "ALL":
            target_risk = riskLevel.upper()
            result = [m for m in result if m.get("riskLevel") == target_risk]

        # Filter plan
        if planType and planType.upper() != "ALL":
            target_plan = planType.lower()
            result = [m for m in result if str(m.get("Plan_Type", "")).lower() == target_plan]

        # Sort
        def sort_key(item):
            val = item.get(sortBy)
            if val is None:
                val = 0
            if isinstance(val, str):
                return val.lower()
            return val

        reverse = (sortOrder.lower() == "desc")
        result.sort(key=sort_key, reverse=reverse)

        # Pagination
        page_num = max(1, page)
        limit_num = max(1, limit)
        total_count = len(result)
        start_index = (page_num - 1) * limit_num
        paginated_members = result[start_index : start_index + limit_num]

        return {
            "totalCount": total_count,
            "page": page_num,
            "limit": limit_num,
            "totalPages": math.ceil(total_count / limit_num) if total_count > 0 else 1,
            "members": paginated_members,
        }

    async def get_member_by_id(self, member_id: str) -> Dict[str, Any]:
        member = DataService.get_member_by_id(member_id)
        if not member:
            raise HTTPException(status_code=404, detail=f"Member {member_id} not found")

        members, schema = DataService.load_dataset()
        processed = self.churn_service.get_processed_dataset()

        pred_res = self.churn_service.predict_member(member)
        probability = pred_res["probability"]
        risk_level = pred_res["riskLevel"]
        prediction = pred_res["prediction"]

        top_drivers = (
            ExplainabilityService.get_member_drivers(
                member, schema, processed, self.churn_service, top_n=5
            )
            if processed else []
        )

        recommendations = RetentionService.generate_recommendations(
            member, risk_level, probability, top_drivers
        )

        # Calculate post-retention churn probability using diminishing returns
        reduction_map = {"HIGH": 0.15, "MEDIUM": 0.10, "LOW": 0.05}
        remaining = probability
        for i, rec in enumerate(recommendations):
            factor = reduction_map.get(rec.get("priority", "LOW"), 0.05)
            remaining *= (1 - factor * (0.7 ** i))
        post_retention_probability = round(max(remaining, 0.01), 4)

        ai_explanation = await GeminiService.generate_member_explanation(
            member_id=str(member.get(schema["idColumn"], "")),
            plan_type=str(member.get("Plan_Type") or "Plan"),
            churn_probability=probability,
            risk_level=risk_level,
            top_drivers=top_drivers,
            approved_actions=recommendations
        )

        return {
            "member": member,
            "churnProbability": probability,
            "riskLevel": risk_level,
            "prediction": prediction,
            "topDrivers": top_drivers,
            "recommendations": recommendations,
            "aiExplanation": ai_explanation,
            "postRetentionProbability": post_retention_probability,
        }
